llvm/lib/Target/RISCV/GISel/RISCVRegisterBanks.td

```diff
@@ -14,3 +14,15 @@ def GPRRegBank : RegisterBank<"GPRB", [GPR]>;
 
 /// Floating Point Registers: F.
 def FPRRegBank : RegisterBank<"FPRB", [FPR64]>;
+
+/// Vector Register Banks:
+def VRRegBank : RegisterBank<"VRB", [VR]>;
```

Comment:
There should only be one vector register bank. A register bank is a collection of registers. It's not one bank for each register class.

Reply:
> There should only be one vector register bank. The register bank should just be the set of all vector registers. It's not one bank for each register class.

Reply:
From the [regbankinfo api](https://releases.llvm.org/9.0.0/docs/GlobalISel.html#api-registerbankinfo) and [gmir-regbank](https://llvm.org/docs/GlobalISel/GMIR.html#gmir-regbank) docs:
```
* Banks: addRegBankCoverage — which register bank covers each register class.
* Register Banks are a means to constrain the register allocator to use a particular register file for a virtual register.
```

The current approach in this patch creates reg bank covers which try to be as small as they can be. For example, if we know that it has a VRM2 class, we choose a VRM2 bank instead of the more broad VR bank. My aim was that we'd constrain the register allocator to only use register that work for M2.

The approach you suggest creates reg bank covers that include the entire register file (the vector registers). What does this achieve if we cannot use every register in the register file anyway?

On a separate, but related note, do you think should VR and VM registers have different banks?

Reply:
Note that description for Register Banks says "particular register file". The vector registers make up a single register file thus should be a single register bank.

> My aim was that we'd constrain the register allocator to only use register that work for M2.

The register allocator doesn't know about register banks. It only knows about register classes. `RISCVInstructionSelector::getRegClassForTypeOnBank` is responsible for converting from RegBank+LLT to register class. LMUL is encoded in the scalable vector type so you should be able to pick the correct LMUL register class from a single bank.

> On a separate, but related note, do you think should VR and VM registers have different banks?

Single register bank.

Reply:
> RISCVInstructionSelector::getRegClassForTypeOnBank is responsible for converting from RegBank+LLT to register class.

It looks like its going from RegClass + LLT -> RegBank. I guess I am confused what the register banks are useful for? 

> The register allocator doesn't know about register banks.

Are you sure this is the case? The docs say `Register Banks are a means to constrain the register allocator...`.

Reply:
> It looks like its going from RegClass + LLT -> RegBank. I guess I am confused what the register banks are useful for?

It should be RegBank+LLT->RegClass like the function name says. It returns 3 values today, `RISCV::GPRRegClass`, `RISCV::FPR32RegClass` or `RISCV::FPR64RegClass`

Reply:
> Are you sure this is the case? The docs say `Register Banks are a means to constrain the register allocator...`.

Once we do instruction selection every register operand or a selected instruction is a virtual register with a register class assigned to it. The register bank and LLT are gone, or at least not used later.